### PR TITLE
[#3368] feat(jdbc-doris): support pre-assign partition for Doris table

### PR DESCRIPTION
**Pull request by xiaozcy**
### What changes were proposed in this pull request?

Support pre-assign partition for Doris table.

### Why are the changes needed?

Fix: #3368 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT & IT

### CODE REVIEW

**Comment by xiaozcy** [comment]
@mchades @zhoukangcn , would you help review this when you are free?

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java [hunk 1]
```diff
@@ -216,11 +220,60 @@ private static void appendPartitionSql(
...
+        ImmutableList.Builder<String> partitions = ImmutableList.builder();
+        for (RangePartition part : assignments) {
+          StringBuilder partitionAssignSqlBuilder = new StringBuilder();
+          partitionAssignSqlBuilder
+              .append(" PARTITION ")
+              .append(BACK_QUOTE)
+              .append(part.name())
+              .append(BACK_QUOTE)
+              .append(" VALUES ");
+          LiteralDTO upper = (LiteralDTO) part.upper();
```
Comment (xiaozcy) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:236]: I wonder why literal values of partitons here are instances of `LiteralDTO`, in my understanding, it seems like they should be instances of `LiteralImpl`, is this a mistake?

Comment (mchades) [reply catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:236]: yes, the mistake is due to the absence of a conversion in DTOConverters. 

Can you prioritize fixing this issue in a separate PR?

Comment (mchades) [reply catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:236]: please rebase the latest main branch and fix this

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java [hunk 2]
```diff
@@ -216,11 +220,60 @@ private static void appendPartitionSql(
           "The partition field must be one of the columns");
 
       String partitionColumn = BACK_QUOTE + rangePartition.fieldName()[0] + BACK_QUOTE;
-      // TODO we currently do not support pre-assign partition when creating range partitioning
-      partitionSqlBuilder.append(partitionColumn).append(") () ");
+      partitionSqlBuilder.append(partitionColumn).append(") ").append(NEW_LINE).append("(");
+
+      RangePartition[] assignments = rangePartition.assignments();
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:225]: Could you split the current `appendPartitionSql` method into several shorter methods? there are too many codes in one method

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/CatalogDorisIT.java [hunk 3]
```diff
@@ -334,14 +358,15 @@ void testDorisTableBasicOperation() {
         };
 
     Map<String, String> properties = createTableProperties();
+    Transform[] partitions = createRangePartition();
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/CatalogDorisIT.java:361]: Please verify that the pre-assigned partitions were successfully created.

**Review by mchades — COMMENTED** [review]
Sorry for the delayed review.

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/operation/TestDorisTableOperations.java [hunk 4]
```diff
@@ -439,7 +445,19 @@ public void testCreateTableWithPartition() {
...
     String rangePartitionTableName = GravitinoITUtils.genRandomName("range_partition_table");
-    Transform[] rangePartition = new Transform[] {Transforms.range(new String[] {col4.name()})};
+    LocalDate today = LocalDate.now();
+    LocalDate tomorrow = today.plusDays(1);
+    Literals.LiteralImpl<LocalDate> todayLiteral = Literals.dateLiteral(today);
+    Literals.LiteralImpl<LocalDate> tomorrowLiteral = Literals.dateLiteral(tomorrow);
+    RangePartition rangePartition1 = Partitions.range("p1", todayLiteral, Literals.NULL, null);
+    RangePartition rangePartition2 = Partitions.range("p2", tomorrowLiteral, todayLiteral, null);
+    RangePartition rangePartition3 = Partitions.range("p3", Literals.NULL, tomorrowLiteral, null);
+    Transform[] rangePartition =
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/operation/TestDorisTableOperations.java:455]: The Transform needs to be verified after creation

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java (hunk 2 above)
Comment (xiaozcy) [reply catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java]: This is already fixed.

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/operation/TestDorisTableOperations.java (hunk 4 above)
Comment (mchades) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/operation/TestDorisTableOperations.java:455]: I mean we should verify that the pre-assigned partitions were successfully created in Doris

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/CatalogDorisIT.java (hunk 3 above)
Comment (mchades) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/CatalogDorisIT.java:360]: Okay, you can make this PR draft before #3961 merged

Comment (mchades) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/CatalogDorisIT.java:360]: Please resolve the conflicts when you have time, thanks!

**Comment by xiaozcy** [comment]
@mchades, this PR is ready for review, thx.

**Review by mchades — APPROVED** [review]
Overall LGTM. @yuqi1129 Do you want to take another look?

**Comment by yuqi1129** [comment]
> Overall LGTM. @yuqi1129 Do you want to take another look?

I'm okay with the current changes.

**Pull request merged**